pkg/controllers/managementuser/rbac/projectresources_handler.go

```diff
@@ -0,0 +1,215 @@
+package rbac
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/rancher/rancher/pkg/api/steve/projectresources"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	pkgrbac "github.com/rancher/rancher/pkg/rbac"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	projectResourcesSuffix = "-projectresources"
+	enqueueAfterPeriod     = 10 * time.Second
+)
+
+// errNotReady is used when the APIResourceWatcher has not performed a successful discovery yet.
+var errNotReady = errors.New("APIs have not been synced yet")
+
+// getProjectResourcesRules accepts a slice of rules and returns a matching set of applicable rules for the resources.project.cattle.io API.
+// The returned list will only include rules that are 1) not for resource names, 2) not for non-resource URLs, 3) use the "list" or "*" verbs, and are namespaced.
+// It returns early if none of the rules match the first 3 conditions, which means the role binding can be skipped.
+// It uses the APIResourceWatcher to check if a resource is namespaced, because only namespaced resources are stored in the watcher.
+// If the APIResourceWatcher has no contents yet, it is probably just starting up so we return an error to indicate that the function should be retried.
+func (m *manager) getProjectResourcesRules(rules []rbacv1.PolicyRule) ([]rbacv1.PolicyRule, error) {
+	validRules := []rbacv1.PolicyRule{}
+	// check that at least one of the rules probably needs to be mirrored before consulting the APIResourceWatcher,
+	// most likely we can return early success and the calling controller can move on from the object.
+	for _, rule := range rules {
+		if len(rule.ResourceNames) != 0 {
+			continue
+		}
+		if len(rule.NonResourceURLs) != 0 {
+			continue
+		}
+		for _, verb := range rule.Verbs {
+			if verb == "*" || verb == "list" {
+				validRules = append(validRules, rule)
+			}
+		}
+	}
+	if len(validRules) == 0 {
+		return []rbacv1.PolicyRule{}, nil
+	}
+	apis := m.apis.List()
+	if len(apis) == 0 {
+		return []rbacv1.PolicyRule{}, errNotReady
+	}
+	projectRules := []rbacv1.PolicyRule{
+		{
+			APIGroups: []string{projectresources.Group},
+			Verbs:     []string{"list"},
+		},
+	}
+	for _, rule := range validRules {
+		for _, apiGroup := range rule.APIGroups {
+			for _, resource := range rule.Resources {
+				_, ok := m.apis.Get(resource, apiGroup)
+				if ok {
+					if apiGroup != "" {
+						resource = apiGroup + "." + resource
```

Comment:
First, it's unlikely (though not impossible) that we'd run into a resource like that. There's no built-in resource type like that, and you can't create a CRD with a `.` in the name (I can't find that documented, but if you try to do it you'll run into a validation message from the server). It's possible this could happen if someone is doing what we're doing, creating an APIService with such a name and group. You'd have to be an admin of the cluster to do this, so they'd be shooting themselves in the foot.

Second, this concatenation is just about finding the resource in the schema, which is already registered in 
https://github.com/rancher/rancher/pull/40155/files#diff-3b10d3f41f25da075a1775137aa0f88a76d011e728a97f248a4136daa81d8584R156 and comes from steve's schema pattern https://github.com/rancher/steve/blob/e6a8019546c176b440922f746e0bc9670b4f9c40/pkg/schema/converter/k8stonorman.go#L31. If overlapping groups is a problem, then the first problem is steve's schema registration, then the new API's schema registration, and then this rbac.

Reply:
Thanks for explaining.